gprofiler/profilers/java.py

```diff
@@ -305,9 +312,10 @@ def _get_stop_cmd(self, with_output: bool) -> List[str]:
         ap_params.append(f"log={self._log_path_process}")
         return self._get_base_cmd() + [",".join(ap_params)]
 
-    def _run_async_profiler(self, cmd: List[str]) -> None:
+    def _run_async_profiler(self, cmd: List[str], timeout: int) -> None:
         try:
-            run_process(cmd)
+            # kill jattach with SIGTERM if it hangs. it will go down
+            run_process(cmd, stop_event=self._stop_event, timeout=timeout, kill_signal=signal.SIGTERM)
```

Comment:
There's logs cleanup code in the `except CalledProcessError` clause, do we want this as well?

Reply:
You mean, the code checking for `_log_path_host`? No reason not to have it (at this point, we don't know if the blocker is jattach itself, or something in AP).

Reply:
Or you meant - we want that code to run for timeouts? Yes, we do

Reply:
After a bit of thinking: not needed - see https://github.com/Granulate/gprofiler/pull/238#discussion_r790306959

Reply:
The error handling for `CalledProcessError` handles AP log as well, when raising `JattachException` the `__str__` also includes AP's log contents, I think it will be useful to have AP's log in the case of timeout as well

Reply:
Hmm. K, I'll make the conversion: I'll extend `CalledProcessError` into `CalledProcessTimeout` which will contain the same fields + the timeout.

In case of `ProcessStoppedException` - we don't want anything.

Reply:
Sounds good, but `CalledProcessTimeoutError ` (Error suffix?)

Reply:
Okie

Reply:
Tested with this diff:
```
diff --git a/gprofiler/profilers/python.py b/gprofiler/profilers/python.py
index 6ca2c2c..db15c53 100644
--- a/gprofiler/profilers/python.py
+++ b/gprofiler/profilers/python.py
@@ -85,7 +85,7 @@ def _add_versions_to_stacks(
 class PySpyProfiler(ProcessProfilerBase):
     MAX_FREQUENCY = 50
     _BLACKLISTED_PYTHON_PROCS = ["unattended-upgrades", "networkd-dispatcher", "supervisord", "tuned"]
-    _EXTRA_TIMEOUT = 10  # give py-spy some seconds to run (added to the duration)
+    _EXTRA_TIMEOUT = -20  # give py-spy some seconds to run (added to the duration)
 
     def __init__(
         self,
```

we now get the timed out log, plus the full stderr stdout etc:
```
[2022-01-24 10:22:34,414] ERROR: gprofiler.profilers.python: Profiling with py-spy timed out on process 1750271
[2022-01-24 10:22:34,415] ERROR: gprofiler.profilers.profiler_base: PySpyProfiler: failed to profile process 1750271
Traceback (most recent call last):
  File "/app/gprofiler/profilers/profiler_base.py", line 111, in snapshot
    result = future.result()
  File "/usr/lib/python3.8/concurrent/futures/_base.py", line 437, in result
    return self.__get_result()
  File "/usr/lib/python3.8/concurrent/futures/_base.py", line 389, in __get_result
    raise self._exception
  File "/usr/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/app/gprofiler/profilers/python.py", line 133, in _profile_process
    run_process(
  File "/app/gprofiler/utils.py", line 258, in run_process
    raise reraise_exc
gprofiler.exceptions.CalledProcessTimeoutError: Timed out after 5 seconds
Command '['/app/gprofiler/resources/python/py-spy', 'record', '-r', '50', '-d', '25', '--nonblocking', '--format', 'raw', '-F', '--gil', '--output', '/tmp/gprofiler_tmp/tmp0hw95d08/pyspy.kiUkeeSWwTKiqubZ.1750271.col', '-p', '1750271', '--full-filenames']' died with <Signals.SIGKILL: 9>.
stdout: b'py-spy> Sampling process 50 times a second for 25 seconds. Press Control-C to exit.\n\n'
stderr: b''
```